homeassistant/helpers/storage.py

```diff
@@ -109,8 +111,12 @@ async def async_load(self) -> dict | list | None:
 
     async def _async_load(self):
         """Load the data and ensure the task is removed."""
+        semaphore = self.hass.data.setdefault(
+            STORAGE_SEMAPHORE, asyncio.Semaphore(MAX_LOAD_CONCURRENTLY)
```

Comment:
Does it matter that a new, most of the time unused, semaphore is created every time this function is called?

Reply:
What am I missing? It should be the same Semaphore object every time.

I tested it with: `_LOGGER.warning("STORAGE_SEMAPHORE: %s", semaphore)`

It locks under pressure 2x during startup on most of my test systems:
```
2021-08-28 19:59:38 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:38 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:39 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:39 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:39 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:2]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:40 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:2]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:1]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:1]>
2021-08-28 19:59:42 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:2]>
2021-08-28 19:59:43 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:2]>
2021-08-28 19:59:43 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:45 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:45 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:45 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:2]>
2021-08-28 19:59:46 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:46 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:47 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:2]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:1]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:1]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:2]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:3]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:4]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:5]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:6]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:7]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:8]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:9]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:10]>
2021-08-28 19:59:48 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [locked, waiters:7]>
2021-08-28 19:59:49 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:49 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:3]>
2021-08-28 19:59:51 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
2021-08-28 19:59:51 WARNING (MainThread) [homeassistant.helpers.storage] STORAGE_SEMAPHORE: <asyncio.locks.Semaphore object at 0x7fa412164580 [unlocked, value:4]>
```

Reply:
The default value in the `dict.setdefault` call is created every time even if it's not set in the dict.

Reply:
Yeah that is inefficient. Bad assumption on my part that cpython would optimize it away if the value already was present. Will adjust.